Remove asserts from codebase

charlesastaylor opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Aims to replace all assert statements (excluding in tests) with conditional statement/exceptions as per [#225](https://github.com/Qiskit/qiskit-terra/issues/225)


### Details and comments
I've initially replaced the asserts in the mapper module. Created a WIP PR to make sure I'm on right lines, with the assert replacements as well as with PR procedure.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=1153) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=1153) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=1153) it.</sub>

charlesastaylor commented:
(As requested moved this question from issue tracker to here) 

I'm not quite sure which Exceptions I should be using. For the MapperError the docstring says "Base class for errors raised by mapper module.", so I was fairly confident about using that one but for some of the other exceptions the docstring is less clear and not sure I should necessarily be blindly replacing all asserts in a module with the exception defined in said module. For example with the VisualizationError docstring - "For visualization specific errors." not really sure just because the assert is in the visualization module whether that exception is correct.

1ucian0 commented:
It seems to me that the `asserts` in the visualization module should raise `VisualizationError`. I agree the docstring is maybe not that clear. Let's continue moving asserts and discussing on the case by case basis.

charlesastaylor commented:
Ok thanks, I'll push my changes shortly.

1ucian0 commented:
In d64c476 I "improved" the documentation and I fixed text visualizer to raise `VisualizationError` instead of `Exception`.

charlesastaylor commented:
In [3d0ba34](https://github.com/Qiskit/qiskit-terra/pull/1153/commits/3d0ba340154d2e1bb649ebc0b2d5bcea3846263b) I've replaced the asserts in the qasm module. Made commit to ask, having noticed you changed the docstring for the visualilzation error module. Would it make sense to change the docstring for the qasm error module similarly?

charlesastaylor commented:
Also, separately, regarding commit messages. For the first commit I followed the guidelines in the contributing file but have noticed commit messages in other PRs and the commit message you used don't do so. Have I missed something? (First PR ever, hence all the questions!)

1ucian0 commented:
About the commit messages, don't worry too much. It's important when squash, so the history in master is kept clean. I will take a look at the doc section to clarify that.

1ucian0 commented:
>   Would it make sense to change the docstring for the qasm error module similarly?

Yes. Take a look first to confirm that the current use of the error fits the new description, just in case.

1ucian0 commented:
I will take a look at the functions in `_simulatortools.py`.  I think some of the assert got moved to `_latex.py`. Can you check it?
```
$ grep -rI 'assert ' qiskit --color
qiskit/tools/visualization/_latex.py:                assert len(op['clbits']) == 1 and len(op['qubits']) == 1
qiskit/tools/visualization/_latex.py:                assert False, "bad node data"
qiskit/tools/visualization/_latex.py:                assert len(op['clbits']) == 1 and \
qiskit/tools/visualization/_latex.py:                    assert False, "If controlled measures currently not supported."
qiskit/tools/visualization/_latex.py:                assert False, "bad node data"
qiskit/backends/aer/_simulatortools.py:    assert i1 != i2
qiskit/unroll/_printerbackend.py:        assert size >= 0, "invalid qreg size"
qiskit/unroll/_jsonbackend.py:        assert size >= 0, "invalid qreg size"
```

charlesastaylor commented:
Ah yep. I didn't recheck after merging the upstream changes.

charlesastaylor commented:
I've replaced those asserts you mentioned. I also changed some QISKITError to VisualizationError in _latex.py as seemed right.
Also as and when I noticed them I had been tidying up some of the import sections to import the exceptions from the relevant package rather than from `package._errormodule`. And I also pep8ed some imports that werent seperated out. Hope that's the right the thing to have done.

mtreinish commented:
@charlesastaylor actually importing from the module is a cleaner way to do it. Since it avoids using a class level import. I've been on a background mission to change all the relative and class level imports when touching code that uses class level or relative imports (they can cause weird circular import issues which makes refactors difficult). So I'd rather you not change those around, especially since they're not related to the change being made in this PR. It's not super critical for the `qiskit.tools.visualization` path because you're switching to something already in the import path from `__init__.py` for submodules off of that, but it's more about changing patterns/norms.

charlesastaylor commented:
@mtreinish thanks for the reply. I'll revert the changes I made. I had to create new imports to use the appropriate Exception so to clarify which style should be favoured for the new imports.

Using QISKitError to demonstrate I've seen 4 ways exceptions are imported. 
`from qiskit import _qiskiterror`
`from qiskit._qiskiterror import QISKitError`
`from ._qiskiterror import QISKitError`
`from qiskit import QISKitError` (where the exception is imported in the relevant `__init__.py` file.

I had been favouring the 4th style and used one of the 3rd style when the exception wasn't imported in the `__init__.py` as it seemed to fit with what was already present. Are you saying that the first option is the preferred? Even when the exception is made available in the init file? (Separately to this PR I could help you on your side mission if you like, sounds like the kind of thing my very limited experience might be able to help with!)

mtreinish commented:
@charlesastaylor I definitely prefer the first method just because it's an absolute import which is explicit (show's where we're importing from in the package namespace) and it's a module import. There is technically nothing wrong with using relative imports like the 3rd way in qiskit (because it's only python>=3.5), it's just I've been contributing/using python long enough and many older versions of python didn't handle them correctly so I typically prefer absoulte imports. The bigger problem is class level imports  since they can cause circular imports (see #911 and #1152 where we got bit by this most recently). 

Anyway this isn't super critical right now because we're already importing QISKitError in `__init__.py` for the qiskit package which means we've got that everywhere already (just because it's always imported for anything in the repo). It's more for reducing future work, like if we want to refactor the imports to be cleaner and use __all__ instead of class level imports to define our top level interface. If you're willing to help with that I'll draft a low priority enhancement issue outlining the direction I'd like to see this all move and mention you there.

charlesastaylor commented:
@mtreinish thanks for the clarification. I believe I have reverted the erroneous changes and prefered the absolute module level import on the new imports I introduced. I have left class level imports of some exceptions (further to QISKitError) as they have been used that way already.

If you think said enhancement would be suitable to warrant a "good first issue" label then I'd be happy for there to be something else I can have a go at!

charlesastaylor commented:
Having noticed [#601](https://github.com/Qiskit/qiskit-terra/issues/601), some of the changes in this PR conflict with what is suggested in that issue.

1ucian0 commented:
Don't worry about #601 for now. One step at the time :)

Please remove the `[WIP]` and squash the commits with a good commit message. If you don't feel confident with squashing (rebasing is always something tricky), just paste the message here as a comment and I can do the rest.

charlesastaylor commented:
Will do tomorrow

charlesastaylor commented:
Thanks. Rebased and squashed.

About the commit message, in the contributing guide it says to reference issue number in first line but looking through log seems like most commits reference the PR number in first line and then an issue number like "Fixes #X" later on. I went for the later as seemed to fit with other messages.

mtreinish commented on qiskit/unrollers/_circuitbackend.py:
```diff
@@ -10,8 +10,7 @@
 from qiskit import QuantumCircuit
-from ._backenderror import BackendError
-from ._unrollerbackend import UnrollerBackend
+from qiskit.unroll import BackendError, UnrollerBackend
```
This looks like a rebase issue. The imports here should be either what was already there in L13-L14 or changed to be an absolute import from `qiskit.unrollers`. Importing from `qiskit.unroll` like this will cause pylint to complain about a circular import because the import path looks like `qiskit.unrollers._circuit_backend` -> `qiskit.unroll` -> `qiskit.unrollers._circuit_backend`.

mtreinish commented on qiskit/unrollers/_jsonbackend.py:
```diff
@@ -34,9 +34,8 @@
-from qiskit.unrollers._unrollerbackend import UnrollerBackend
-from qiskit import QISKitError
+
+from qiskit.unroll import BackendError, UnrollerBackend
```
This looks like a rebase issue. The imports here should be either what was already there in L37-L39 or changed to be an absolute import `from qiskit.unrollers`. Importing from qiskit.unroll like this will cause pylint to complain about a circular import because the import path looks like `qiskit.unrollers._jsonbackend` -> `qiskit.unroll` -> `qiskit.unrollers._jsonbackend`.

mtreinish commented on qiskit/unrollers/_printerbackend.py:
```diff
@@ -8,8 +8,8 @@
-from ._backenderror import BackendError
-from ._unrollerbackend import UnrollerBackend
+
+from qiskit.unroll import BackendError, UnrollerBackend
```
This looks like a rebase issue. The imports here should be either what was already there in L11-L12 or changed to be an absolute import `from qiskit.unrollers`.

charlesastaylor commented:
@mtreinish Hi, yep I had noticed my mistake myself. Some of the tests were failing then noticed the change of package name. I've changed all the imports now back to what should be. However some different tests are still failing now instead.

mtreinish commented:
@charlesastaylor that error looks like it's caused by the introduction of a new test requirement (which was added in #1154 ) You'll likely only hit that locally and it can be fixed by just reinstalling the requirements in requirements-dev.txt on the environment you're running the tests in (or just run `pip install jupyter ipywidgets`, which are the new requirements).

1ucian0 commented:
Congrats @charlesastaylor! your first PR is merged!

charlesastaylor commented:
Thanks! Time to get looking for something else I can have a go at.
